core/src/main/java/org/elasticsearch/index/query/BoolQueryParser.java

```diff
@@ -105,14 +100,13 @@ public QueryBuilder fromXContent(QueryParseContext parseContext) throws IOExcept
                     switch (currentFieldName) {
                     case "must":
                         query = parseContext.parseInnerQueryBuilder();
-                        if (query != null) {
-                            mustClauses.add(query);
-                        }
+                        mustClauses.add(query);
                         break;
                     case "should":
                         query = parseContext.parseInnerQueryBuilder();
-                        if (query != null) {
-                            shouldClauses.add(query);
+                        shouldClauses.add(query);
+                        // EmptyQueryBuilder does not add lucene query later, skip setting minuminShouldMatch
+                        if (query != EmptyQueryBuilder.PROTOTYPE) {
```

Comment:
ok lets get this PR in and address this problem separately, I think we overlooked this when migrating the bool query. This check should be moved to the toQuery method. We should start looking into dividing the context into a query parse context and a to query context I think